frame/utility/src/lib.rs

```diff
@@ -459,6 +483,66 @@ pub mod pallet {
 			let base_weight = T::WeightInfo::batch(calls_len as u32);
 			Ok(Some(base_weight + weight).into())
 		}
+
+		/// Forwards a pre-signed function call.
+		///
+		/// This method allows to dispatch the pre-signed call by relayer on user's behalf
+		///
+		/// - `forwarded_call`: A struct that consists:
+		/// 	- `call`: the call to be forwarded
+		/// 	- `nonce`: a nonce of the account that signed the call
+		/// 	- `deadline`: a block number this call is valid until
+		/// 	- `signer`: an account that signed the `call`
+		/// 	- `whitelisted_forwarder`: who could forward the call (optionally)
+		/// - `signature`: a signature of the `forwarded_call` struct
+		#[pallet::weight({
+			let dispatch_info = forwarded_call.call.get_dispatch_info();
+			(
+				T::WeightInfo::submit_call_on_behalf_of()
+					.saturating_add(dispatch_info.weight),
+				dispatch_info.class,
+			)
+		})]
+		pub fn submit_call_on_behalf_of(
+			origin: OriginFor<T>,
+			forwarded_call: ForwardedCallOf<T>,
+			signature: MultiSignature,
+		) -> DispatchResultWithPostInfo {
+			let sender = ensure_signed(origin.clone())?;
+			ensure!(forwarded_call.verify(&signature), Error::<T>::InvalidSignature);
+
+			let ForwardedCall { signer: who, deadline, call, whitelisted_forwarder, nonce } =
+				forwarded_call;
+
+			if let Some(forwarder) = whitelisted_forwarder {
+				ensure!(forwarder == sender, Error::<T>::WrongForwarder);
+			}
+
+			// validate deadline
+			let now = frame_system::Pallet::<T>::block_number();
+			ensure!(deadline >= now, Error::<T>::AuthorizationExpired);
+
+			// convert AccountId32 to AccountId
+			let who = who.into_account();
+			let who = T::AccountId::decode(&mut AccountId32::as_ref(&who))
+				.map_err(|_| Error::<T>::ErrorConvertingToAccountId)?;
+
+			// validate nonce
+			let account = Account::<T>::get(&who);
+			if nonce != account.nonce {
+				return Err(Error::<T>::InvalidNonce.into())
+			}
+
+			// emit event and dispatch call
+			let call_hash = blake2_256(&Encode::encode(&call));
```

Comment:
```suggestion
			let call_hash = T::Hash::hash(&Encode::encode(&call));
```

Why not the block hash?

Reply:
And you also encode the call two times. Once here and then inside the `verify` function.

Reply:
Inside the `verify` function I encode the whole `ForwardedCall` struct, while here it encodes only the inner `call` param